### ISSUE TITLE
sigsegv on 2.5.2

### ISSUE THREAD

**Issue by schmurfy** [rubinius/rubinius#3330]
I give rubinius a try regularly and just tried again with 2.5.2, while running my test suite I get a sigsegv pretty quickly which seems to be related to eventmachine but the trace shows that the error originates from rubinius internals.

ruby and lldb backtraces are available here: https://gist.github.com/schmurfy/509c4efc0e840fccb912

I am running on Mac OS X 10.9.4
I currently don't have any simple way to make a simple test case to reproduce this crash since I don't really know where it happens, this is the test suite for a fairly big project.

The crash seems to happen around a call to the database using the mysql2 gem via activerecord though.
I can try to provide more data if I know what/where to look.

**Comment by schmurfy** [comment]
I just tested with master and I have the same crash.

**Comment by schmurfy** [comment]
I tried to reproduce the crash with a minimal test case but for now I failed :(
The strange thing is that now rubinius just stays there taking cpu until I control+c it and then it shows me the sigsegv error, can I connect to a running process with the debugger to see what it is currently doing ?

**Comment by YorickPeterse** [comment]
@schmurfy Using GDB is probably the easiest way here since it also gives insight in what's happening in C++. A small guide on using GDB can be found here: http://rubini.us/doc/en/how-to/obtaining-gdb-backtraces/

**Comment by brixen** [comment]
@schmurfy is there a way we could run your tests? To rule out mysql2, can you run with sqlite3? Also, the only thing I see in the stack trace that isn't Rubinius appears to be EventMachine. Do you know what version you're using?

**Comment by schmurfy** [comment]
I tried to reproduce the crash with a minimal application but failed and can't share this one :\
I am using eventmachine-le 1.1.7

I can try gathering more data but I am out of ideas, I tested again a few days ago and the crash still happened with master.

**Comment by brixen** [comment]
@schmurfy would you be able to build with `./configure --debug-build`, run under the debugger, and leave the session running and gist `thread backtrace -c 15 all` when it faults?

**Comment by schmurfy** [comment]
sure, I will test it this week when I have a moment.

**Comment by brixen** [comment]
@schmurfy awesome, thanks.

**Comment by schmurfy** [comment]
I just did the test with master branch and the segfault appears to be gone, instead I now have a SystemStackError error raised (but with a rather small backtrace which is weird), I am trying to figure out where it comes from and will report here if I gather useful data on whether there is still a rubinius issue or not.

**Comment by schmurfy** [comment]
I did the test with 2.5.5 to see if there was an issue with master but the behavior is the same as previously, the segfault is gone and I still have no clue where the SystemStackError is coming from but at least it is no longer crashing the interpreter xD

I think this issue can be closed since the original issue is no longer true in the latest release, I will open another one if I can gather anything useful on this stack error (and if this is rubinius related)).

**Comment by brixen** [comment]
@schmurfy thanks for the update. If you want to paste the SystemStackError backtrace, we can take a look.

I'm curious what parts of EM you're using in your app. We are looking at making an API-compatible EM replacement that is pure Ruby and I'd like to focus on the parts that people actually use. Any help with your use cases would be great.

**Comment by jc00ke** [comment]
Reminds me of oldmoe's [Reactor](https://github.com/oldmoe/reactor) :wink:

**Comment by schmurfy** [comment]
The applications I am testing rubinius with is a the server side of a client-server application, we are using eventmachine to listen on udp and tcp connections and answer requests.

The reason we use eventmachine and not just the existing ruby socket api was initally for performance (we started with ruby 1.8 and having the reactor written in C was an interesting option), the server was also designed to handle every request asynchronously using fibers which in our case works great until now, eventmachine works well with this model.

One of the reason I try rubinius on a regular basis is that I hope to be able to turn this application into a multi threaded server at some point without any GIL.

**Comment by brixen** [comment]
@schmurfy good to know, thanks! Have you experimented with just moving the work items from EM to a queue and thread pool?

**Comment by schmurfy** [comment]
EM is only used as a reactor, we don't really use any of its higher level constructs, basically it handles the sockets read/write and then we dispatch the work on a pool of fibers.

Before we started this app we did some tests first with a purely asynchronous approach which was a major pain, then tried to use threads to keep the context and avoid the callback hell, fibers came last (MRI 1.9 was brand new) and were adopted without much opposition. Performances were similar if not better with the single threaded multi fibers approach and it (mostly) threw concurrency issues out of the window since we know when the switch between fibers can/will happen.

The projects I am currently looking at are concurrent-ruby (for handling concurrency) and nio4r as an EM replacement (a reactor, lower level than eventmachine), eventmachine was good at the time but I don't like its current state and would really like to move away from it, here are some of the reasons:
- issues piling up with not much activity, or more exactly activity on the commit side but unrelated to the issues (that's why eventmachine-le was created so I don't seem to be the only one feeling that way)
- single reactor model: I don't see any reason for that limitation and in our case multiple reactors would be useful.
- eventmachine is "in theory" able to run anywhere thanks to multiple implmentations (C, ruby, java) but in practice they are not compatible with each other so they are just a burden and a trap (different state of their api) and some are more maintained than others, in practice for us we had to include a raise to stop everything if anything but the the C extension was chosen by the runtime (in some case it felt  back to the ruby version which simply does not work for us).
I tried jruby briefly but eventmachine shot down this attempt rather quickly for the reason stated above.

nio4r looks like the most promising replacement for the network part of EM for me, I even started working on a gem (which is there on my github: nio4r-em) a few years ago to provide the EM api above nio4r to ease transition/testing but I don't think I ever finished it although it should not be that far.

I am not sure an EM like api in rubinius itself would be a good thing, MRI already have its standard library which is mostly a dead weight in my opinion, I don't say that the libraries in there are all bad but it's just that I will always prefer using the latest version (the gem) instead of an old version shipped with MRI. If a reactor api is inside rubinius I fear the same thing would happen, I think having it as a separate gem is a better approach.
